### PR TITLE
Refactor writing of bib files

### PR DESCRIPTION
**Pull request by tobiasdiez**
The main aim of this PR is to simplify the implementation of https://github.com/JabRef/jabref/pull/1451 (sql export) and other exporters.

- Abstract class BibDatabaseWriter, which controls the generic things of
writing a database (i.e. sorting entries, applying save actions, some
basic conversations, ...)
- Derived class BibtexDatabaseWriter, which only contains logic how to
actually write to in BibTeX format (should be simple to write similar
classes for other export formats like ris)
- Make SaveSession abstract and introduce two implementations which
write to a temporary file (FileSaveSession) or two a string/buffer
(StringSaveSession)
- Move code related to lock files to FileBasedLock

--

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [ ] Screenshots added (for bigger UI changes)

### CODE REVIEW

**Comment by tobiasdiez** [comment]
Any comments? Can I merge this in (of course after fixing compile errors)?

File: src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java [hunk 1]
```diff
@@ -33,37 +34,50 @@
...
+
+    public interface SaveSessionFactory<E extends SaveSession> {
+        E createSaveSession(Charset encoding, Boolean makeBackup) throws SaveException;
+    }
+
+    private static List<FieldChange> applySaveActions(List<BibEntry> toChange, MetaData metaData) {
+        List<FieldChange> changes = new ArrayList<>();
+
+        Optional<FieldFormatterCleanups> saveActions = metaData.getSaveActions();
+        if (saveActions.isPresent()) {
```
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java]: Mabye use the lambda version: ifPresent

File: src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java [hunk 2]
```diff
@@ -208,162 +197,60 @@ public SaveSession savePartOfDatabase(BibDatabaseContext bibDatabaseContext, Lis
                         entryType -> typesToWrite.put(entryType.getName(), entryType));
```
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java]: I think this is double code: First you check ! present and then you check the same with ifPresent..

File: src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java [hunk 3]
```diff
@@ -208,162 +197,60 @@ public SaveSession savePartOfDatabase(BibDatabaseContext bibDatabaseContext, Lis
...
-        }
-    }
+    protected abstract void writeEpilogue(String epilogue) throws SaveException;
 
     /**
      * Writes all data to the specified writer, using each object's toString() method.
      */
-    private void writeMetaData(Writer out, MetaData metaData) throws IOException {
+    protected void writeMetaData(MetaData metaData) throws SaveException {
         if (metaData == null) {
```
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java]: Maybe use optional?

File: src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java [hunk 4]
```diff
@@ -376,32 +263,55 @@ private void writeStrings(Writer fw, BibDatabase database, Boolean reformatFile)
...
+        // Then we go through the string looking for references to other strings. If we find references
+        // to strings that we will write, but still haven't, we write those before proceeding. This ensures
+        // that the string order will be acceptable for BibTeX.
+        String content = bibtexString.getContent();
+        Matcher m;
+        while ((m = REFERENCE_PATTERN.matcher(content)).find()) {
+            String foundLabel = m.group(1);
+            int restIndex = content.indexOf(foundLabel) + foundLabel.length();
+            content = content.substring(restIndex);
+            Object referred = remaining.get(foundLabel.substring(1, foundLabel.length() - 1));
```
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java]: First: Why Object? And second: Better use getOrDefault, avoids the null check

File: src/main/java/net/sf/jabref/exporter/BibtexDatabaseWriter.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    protected void writeEntry(BibEntry entry, BibDatabaseMode mode, Boolean isReformatFile) throws SaveException {
+        BibEntryWriter bibtexEntryWriter = new BibEntryWriter(new LatexFieldFormatter(), true);
+        try {
+            bibtexEntryWriter.write(entry, getWriter(), mode, isReformatFile);
+        } catch (IOException e) {
+            throw new SaveException(e, entry);
+        }
+    }
+
+    private Writer getWriter() {
```
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/exporter/BibtexDatabaseWriter.java]: I would suggest making that a field, e.g. Writer writer and use that everywhere instead of getWriter()
Improves readability

File: src/main/java/net/sf/jabref/exporter/ExportFormat.java [hunk 6]
```diff
@@ -190,15 +190,15 @@ public void performExport(final BibDatabaseContext databaseContext, final String
         SaveSession ss = null;
```
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/exporter/ExportFormat.java]: Use nio Path method above.

File: src/main/java/net/sf/jabref/exporter/FileSaveSession.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    private static Path createTemporaryFile() throws SaveException {
+        try {
+            return Files.createTempFile(FileSaveSession.TEMP_PREFIX, FileSaveSession.TEMP_SUFFIX);
+        } catch (IOException e) {
+            throw new SaveException(e);
+        }
+    }
+
+    @Override
+    public void commit(File file) throws SaveException {
```
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/exporter/FileSaveSession.java]: Use nio methods 
e.g. File.exist
and  sth like Path.get(path, name, backup Extension)

File: src/main/java/net/sf/jabref/exporter/FileSaveSession.java [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+                        if (!FileBasedLock.waitForFileLock(file, 10)) {
+                            throw SaveException.FILE_LOCKED;
+                        }
+                    }
+                } catch (IOException ex) {
+                    LOGGER.error("Error when creating lock file.", ex);
+                }
+            }
+
+            FileUtil.copyFile(temporaryFile.toFile(), file, true);
```
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/exporter/FileSaveSession.java]: Use Files.copy...

File: src/main/java/net/sf/jabref/logic/util/io/FileBasedLock.java [hunk 9]
```diff
@@ -54,16 +59,17 @@ public static boolean hasLockFile(File file) {
      * @return the last modified time if lock file exists, -1 otherwise.
      */
     public static long getLockFileTimeStamp(File file) {
-        File lock = new File(file.getPath() + SaveSession.LOCKFILE_SUFFIX);
+        File lock = new File(file.getPath() + LOCKFILE_SUFFIX);
```
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/logic/util/io/FileBasedLock.java]: Paths.get

File: src/main/java/net/sf/jabref/logic/util/io/FileBasedLock.java [hunk 10]
```diff
@@ -72,4 +78,25 @@ public static boolean deleteLockFile(File file) {
...
     }
+
+    /**
+     * Check if a lock file exists, and create it if it doesn't.
+     *
+     * @return true if the lock file already existed
+     * @throws IOException if something happens during creation.
+     */
+    public static boolean createLockFile(File file) throws IOException {
+        File lock = new File(file.getPath() + LOCKFILE_SUFFIX);
```
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/logic/util/io/FileBasedLock.java]: same as above, Paths.,get

File: src/main/java/net/sf/jabref/logic/util/io/FileBasedLock.java [hunk 11]
```diff
@@ -72,4 +78,25 @@ public static boolean deleteLockFile(File file) {
...
+        if (lock.exists()) {
+            return true;
+        }
+        try (FileOutputStream out = new FileOutputStream(lock)) {
+            out.write(0);
+            out.close();
```
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/logic/util/io/FileBasedLock.java]: Is the close necessary? When you use a try with resource (Autocloseable), normally all Streams are closed and freed

File: src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java [hunk 12]
```diff
@@ -653,4 +653,7 @@ public static String repeat(int n, char c) {
 
     }
 
+    public static boolean isNullOrEmpty(String toTest) {
+        return (toTest == null || toTest.isEmpty());
```
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/logic/util/strings/StringUtil.java]: You can use the apache commons StringUtil method for this

File: src/test/java/net/sf/jabref/exporter/BibtexDatabaseWriterTest.java [hunk 13]
```diff
@@ -40,186 +37,175 @@
...
     @Test(expected = NullPointerException.class)
-    public void writeWithNullPreferencesThrowsException() throws IOException {
-        databaseWriter.writePartOfDatabase(mock(Writer.class), bibtexContext, Collections.emptyList(), null);
+    public void writeWithNullPreferencesThrowsException() throws Exception {
+        databaseWriter.savePartOfDatabase(bibtexContext, Collections.emptyList(), null);
     }
 
     @Test
-    public void writeEncoding() throws IOException {
+    public void writeEncoding() throws Exception {
```
Comment (Siedlerchr) [inline src/test/java/net/sf/jabref/exporter/BibtexDatabaseWriterTest.java]: Why did you remove all specific Exceptions?
Better let it throw the most specific exception. Otherwise stacktrace can be hard to find.

**Comment by Siedlerchr** [comment]
Overall looks good, just some small remarks from my side :+1:

File: src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java (hunk 1 above)
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java]: Yeah, but I find the old-style version more readable.

File: src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java (hunk 2 above)
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java]: With a small difference: the first check is for `StandardType` (i.e. if it's a JabRef default type) and the second one a `KnownType` (i.e. standard or user-specfied).

File: src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java (hunk 3 above)
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java]: `requireNonNull` is even more appropriate, I think.

File: src/test/java/net/sf/jabref/exporter/BibtexDatabaseWriterTest.java (hunk 13 above)
Comment (tobiasdiez) [inline src/test/java/net/sf/jabref/exporter/BibtexDatabaseWriterTest.java]: Makes the tests more invariant under changes of the code-under-test. (and is default template for IntellliJ)

**Comment by tobiasdiez** [comment]
@Siedlerchr Thanks for you feedback. Incorporated it and rebased.

File: src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java [hunk 14]
```diff
@@ -90,11 +104,11 @@
     }
 
     /*
-     * We have begun to use getSortedEntries() for both database save operations
-     * and non-database save operations.  In a non-database save operation
-     * (such as the exportDatabase call), we do not wish to use the
-     * global preference of saving in standard order.
-     */
+         * We have begun to use getSortedEntries() for both database save operations
```
Comment (koppor) [inline src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java]: indent?

File: src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java (hunk 1 above)
Comment (simonharrer) [inline src/main/java/net/sf/jabref/exporter/BibDatabaseWriter.java]: Optional only make sense if one uses ifPresent or map/flatMap. Otherwise, it does not really help that much, from my perspective, as one still requires an if check and a get, making it even more complicated to use instead of a null check and just using it.

**Comment by simonharrer** [comment]
@tobiasdiez after rebase and successful tests, this can be merged as well.
